### DISCUSSION TITLE
Welcome to ramp-code Discussions!

### DISCUSSION THREAD

**Discussion by cjohnstonconsulting** [devglobalpartners/ramp-code#4] [Announcements]
## 👋 Welcome!

We’re using Discussions as a point of support for the entire ramp user community. We hope that you:

  * Ask questions about the ramp codebase (and possible extensions and future projects) 
  * Ask for help with ramp use cases
  * Get connected with the ramp team and other ramp users!

Please introduce yourself below!

**Comment by cjohnstonconsulting** [comment]
Hello everyone and welcome,

I'm Carolyn Johnston of [Johnston Consulting Services, LLC](https://carolynpjohnston.wordpress.com/), a small (okay it's just me) business specializing in data science for geospatial applications. I teamed up with DevGlobal Partners to do the ML and codebase development for the original ramp project. I'm available to support ramp users and code contributors, and look forward to contributing to discussions and helping solve issues! ([LinkedIn](https://www.linkedin.com/in/carolyn-johnston-5761956/))

**Comment by omranlm** [comment]
Hello all,
This is Omran NAJJAR, [AI and Advanced Data Engineer](https://www.hotosm.org/people/omran-najjar/) with Humanitarian OpenStreetMap Team (HOT), collaborated as RAMP advisory member during the implementation of RAMP. Happy to contribute and support!

**Reply by cjohnstonconsulting** [reply]
Welcome, Omran, so glad to have you join us here!

**Comment by seamusgeraty** [comment]
Hi Everybody!

I'm Seamus Geraty, a junior consultant with DevGlobal Partners, and I worked with Carolyn and the team to develop the original Ramp model. I spent a lot of time working with the Ramp training data, tools and documentation and am eager to help out where I can. Don't hesitate to reach out!

**Reply by cjohnstonconsulting** [reply]
Welcome, Seamus!

**Comment by shaimaahegazy** [comment]
Hi everyone, I'm experimenting using the RAMP model on a problem that we're solving, and I wanted to ask about the "get_model_iou.py" script mentioned [here](https://rampml.global/ramp-scripts/). I don't seem to find this script in the directory, any clues? Thanks a lot!

**Reply by seamusgeraty** [reply]
Noted. Looking into it and will follow up soon

**Reply by seamusgeraty** [reply]
Hi Shaimaa, 

It seems that the "get_model_iou.py" entry in the documentation is outdated and the script has since been removed. The functionality of that script is now captured in "calculate_accuracy_iou.py", which you should be able to access in the directory. Hope this helps, if not let us know and we can dig into it more. 

Thanks for bringing it to our attention, we will update the documentation and remove the inaccuracy.

**Reply by vrangayyan6** [reply]
Team, you have provided [Final Model Checkpoint](https://radiantearth.github.io/model_ramp_baseline/data/input/checkpoint.tf) for model_ramp_baseline_v1.  tried to use it for fine tuning with dataset from HOT in [train_ramp_model_on_colab.ipynb](https://github.com/devglobalpartners/ramp-code/blob/main/colab/train_ramp_model_on_colab.ipynb), with attached config file [ramp_tf_HOTOSM_config_1.txt](https://github.com/devglobalpartners/ramp-code/files/9883852/ramp_tf_HOTOSM_config_1.txt) and getting error on executing following cell -

-----------------------------------------
```
the_model = None

if cfg["saved_model"]["use_saved_model"]:

    
    # load (construct) the model
    model_path = Path(RAMP_HOME) / cfg["saved_model"]["saved_model_path"]
    print(f"Model: importing saved model {str(model_path)}")
    the_model = tf.keras.models.load_model(model_path)
    assert the_model is not None, f"the saved model was not constructed: {model_path}"

    if not cfg["saved_model"]["save_optimizer_state"]:
        # If you don't want to save the original state of training, recompile the model.
        the_model.compile(optimizer = optimizer, 
            loss=loss_fn,
            metrics = the_metrics)
```
------------------------------------------------------
OUTPUT

```
Model: importing saved model /content/drive/MyDrive/RAMP_HOME/ramp-data/TRAIN/checkpoint.tf

---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-37-c339319085e8> in <module>
      6     model_path = Path(RAMP_HOME) / cfg["saved_model"]["saved_model_path"]
      7     print(f"Model: importing saved model {str(model_path)}")
----> 8     the_model = tf.keras.models.load_model(model_path)
      9     assert the_model is not None, f"the saved model was not constructed: {model_path}"
     10 

/usr/local/lib/python3.7/dist-packages/keras/utils/traceback_utils.py in error_handler(*args, **kwargs)
     65     except Exception as e:  # pylint: disable=broad-except
     66       filtered_tb = _process_traceback_frames(e.__traceback__)
---> 67       raise e.with_traceback(filtered_tb) from None
     68     finally:
     69       del filtered_tb

/usr/local/lib/python3.7/dist-packages/tensorflow/python/saved_model/loader_impl.py in parse_saved_model(export_dir)
    103       return saved_model
    104     except message.DecodeError as e:
--> 105       raise IOError(f"Cannot parse file {path_to_pb}: {str(e)}.")
    106   elif file_io.file_exists(path_to_pbtxt):
    107     with file_io.FileIO(path_to_pbtxt, "rb") as f:

OSError: Cannot parse file b'/content/drive/MyDrive/RAMP_HOME/ramp-data/TRAIN/checkpoint.tf/saved_model.pb': Error parsing message.
```
-------------------------------------------------

can you please advise?

**Reply by shaimaahegazy** [reply]
> Hi Shaimaa,
> 
> It seems that the "get_model_iou.py" entry in the documentation is outdated and the script has since been removed. The functionality of that script is now captured in "calculate_accuracy_iou.py", which you should be able to access in the directory. Hope this helps, if not let us know and we can dig into it more.
> 
> Thanks for bringing it to our attention, we will update the documentation and remove the inaccuracy.

Thanks a lot, highly appreciating your reply. I was able to use it to get the IoU. Thank you.

**Reply by shaimaahegazy** [reply]
> Hi Shaimaa,
> 
> It seems that the "get_model_iou.py" entry in the documentation is outdated and the script has since been removed. The functionality of that script is now captured in "calculate_accuracy_iou.py", which you should be able to access in the directory. Hope this helps, if not let us know and we can dig into it more.
> 
> Thanks for bringing it to our attention, we will update the documentation and remove the inaccuracy.

Hi Seamus,

I've got another question please. the script provides me with the precision, recall and F1 metrics, is there a way to calculate the mean IoU and get that number?

Thank you!

**Reply by cjohnstonconsulting** [reply]
Hi Shaimaa, 

ramp does not have a script that explicitly calculates the mean IoU, although such a script could easily be written with the functions in the ramp and solaris modules. 

Best, Carolyn

**Reply by shaimaahegazy** [reply]
Sounds great, that's what we did! Thanks a lot Carolyn.

**Reply by cjohnstonconsulting** [reply]
I love it! Would you consider contributing the new script to the ramp codebase?

**Reply by shaimaahegazy** [reply]
> I love it! Would you consider contributing the new script to the ramp codebase?

yes sure! please let me know how to do that.

**Comment by cjohnstonconsulting** [comment]
Hi vrangayyan6,

I'm sorry for the confusion. I have not personally worked (yet) with the RadiantEarth dockerized saved model, but I do know it is packaged to be used for inferencing, rather than as a baseline model in a new training regimen. 

However, Radiant Earth did tell me that there is a workaround that lets you access the baseline model for training.  So I would recommend giving this a try (warning, your mileage may vary and I have not tried this myself yet!):

The [RadiantEarth model README](https://radiantearth.github.io/model_ramp_baseline/) states that their model repository contains a model checkpoint in tensorflow format, which is a directory named 'checkpoint.tf'. Try replacing the cfg["saved_model"]["saved_model_path"] value with the path to this model checkpoint directory. The 'load_model' function should be able to read the model data in that format. 

Please let us know how it goes! If that doesn't work, then we need to figure out a solution that does. :)

And thank you!
Best wishes,
Carolyn

**Reply by vrangayyan6** [reply]
Thanks Carolyn.  I had not installed **git-lfs** and then cloned model_ramp_baseline.  Second time I got this error, then I replaced checkpoint.tf in my Google Drive.  

My team member Shaimaa told me to use below in configuration which worked in finetuning -
"saved_model":{
	"use_saved_model":true,
	"saved_model_path":"ramp-code/ramp/checkpoint.tf",
	"save_optimizer_state":false
},

It may help to add steps to be taken for fine tuning to page https://rampml.global/project-introduction/ > section Training Data, Model Weights – MLHub

**Comment by vrangayyan6** [comment]
Sometimes, after executing the last cell and on clicking the link to Juptyer Lab in [jupyter_lab_on_colab.ipynb](https://github.com/devglobalpartners/ramp-code/blob/main/colab/jupyter_lab_on_colab.ipynb), I am getting error as in attachment

![jupyter_lab_on_colab](https://user-images.githubusercontent.com/18081616/199642686-f128210f-e0ae-47b4-b64b-beb94339cfc5.png)

![colab_jupyter_lab_error](https://user-images.githubusercontent.com/18081616/199642696-bb0af2b9-7dcd-4582-a240-44a5bee32bb6.png)

I have cleared the browser cache and tried different browsers but it did not help.  Do you have any advice? 

Thanks,
Vinod

**Reply by mmschmitz** [reply]
Hi all - Michelle Schmitz here; I work as a Senior Consultant, with Seamus and Carolyn, in the geospatial practice at DevGlobal Partners. (That said, I did not develop the RampML prototype!)

@vrangayyan6  - have you tried [the following instructions](https://stackoverflow.com/questions/55554351/no-web-browser-found-could-not-locate-runnable-browser-jupyter-notebook)? 

If it doesn't seem to work, please reply and we'll see if we can get to the bottom of this!

**Reply by vrangayyan6** [reply]
referring to https://medium.com/@apbetahouse45/how-to-run-jupyter-notebooks-on-remote-server-part-1-ssh-a2be0232c533 

executed
`ssh -NL 80:localhost:9090 root@3.234.18.192`

got error
```
Warning: Permanently added '3.234.18.192' (RSA) to the list of known hosts.
Received disconnect from 3.234.18.192 port 22:10: no tunnel request seen
Disconnected from 3.234.18.192 port 22
```

**Comment by shaimaahegazy** [comment]
Hello, 

I actually have a couple of questions regarding using the RAMP. 

Firstly, If we'd like to integrate it in our library (semantic segmentation library), what's the easiest way to do that? (we mainly need to finetune the model on our input images).

Secondly, can we pass a mosaic to the inference engine rather than passing the individual chips?

Thank you!
